swayfmt/src/utils/language/expr/mod.rs

```diff
@@ -544,46 +550,110 @@ impl Format for Expr {
+                    // the current expr is not being rendered multiline at the
+                    // expr level, then add an indentation to the following
+                    // expression and generate the code
+                    formatter.with_shape(
+                        formatter
+                            .shape
+                            .with_code_line_from(LineStyle::Multiline, ExprKind::Undetermined),
+                        |formatter| -> Result<(), FormatterError> {
+                            formatter.shape.code_line.update_expr_new_line(true);
+
+                            lhs.format(formatted_code, formatter)?;
+                            formatter.indent();
+                            write!(
+                                formatted_code,
+                                "\n{}{} ",
+                                formatter.indent_to_str()?,
+                                double_ampersand_token.span().as_str()
+                            )?;
+                            rhs.format(formatted_code, formatter)?;
+                            formatter.unindent();
+                            Ok(())
+                        },
+                    )?;
+                } else {
+                    lhs.format(formatted_code, formatter)?;
+                    match formatter.shape.code_line.line_style {
+                        LineStyle::Multiline => {
+                            write!(
+                                formatted_code,
+                                "\n{}{} ",
+                                formatter.indent_to_str()?,
+                                double_ampersand_token.span().as_str()
+                            )?;
+                        }
+                        _ => {
+                            write!(
+                                formatted_code,
+                                " {} ",
+                                double_ampersand_token.span().as_str()
+                            )?;
+                        }
                     }
+                    write!(formatted_code, "{}", rhs_code)?;
                 }
-                rhs.format(formatted_code, formatter)?;
             }
             Self::LogicalOr {
                 lhs,
                 double_pipe_token,
                 rhs,
             } => {
-                lhs.format(formatted_code, formatter)?;
-                match formatter.shape.code_line.line_style {
-                    LineStyle::Multiline => {
-                        write!(
-                            formatted_code,
-                            "\n{}{} ",
-                            formatter.indent_to_str()?,
-                            double_pipe_token.span().as_str()
-                        )?;
-                    }
-                    _ => {
-                        write!(formatted_code, " {} ", double_pipe_token.span().as_str())?;
+                let mut rhs_code = FormattedCode::new();
```

Comment:
See my comment on the same code here https://github.com/FuelLabs/sway/pull/5405#discussion_r1434583575